lighthouse/helpers/reports.py

```diff
@@ -241,25 +238,85 @@ def get_cherrypicked_samples(root_sample_ids, plate_barcodes, chunk_size=50000):
         db_connection.close()
 
 
-def get_all_positive_samples(samples):
-
-    logger.debug("Getting all positive samples")
-    # filtering using case insensitive regex to catch "Positive" and "positive"
-    results = samples.find(
-        filter=POSITIVE_SAMPLES_MONGODB_FILTER,
-        projection={
-            "_id": False,
-            FIELD_SOURCE: True,
-            FIELD_PLATE_BARCODE: True,
-            FIELD_ROOT_SAMPLE_ID: True,
-            FIELD_RESULT: True,
-            FIELD_DATE_TESTED: True,
-            FIELD_COORDINATE: True,
+def get_all_positive_samples(samples_collection: Collection) -> DataFrame:
+    """Get all the positive samples from mongo from a specific date.
+
+    Args:
+        samples_collection (Collection): the samples collection.
+
+    Returns:
+        DataFrame: a pandas DataFrame with the positive samples.
+    """
+    logger.debug("Getting all positive samples from a specific date")
+
+    # The projection defines which fields are present in the documents from the output of the mongo
+    # query
+    projection = {
+        "_id": False,
+        FIELD_SOURCE: True,
+        FIELD_PLATE_BARCODE: True,
+        FIELD_ROOT_SAMPLE_ID: True,
+        FIELD_RESULT: True,
+        FIELD_DATE_TESTED: True,
+        FIELD_COORDINATE: True,
+    }
+
+    DATE_EXTRACTED = "extracted_date"
+    DATE_CONVERTED = "converted_date"
+    DATE_ERROR = "date_conversion_error"
+
+    # The pipeline defines stages which execute in sequence
+    pipeline = [
+        # 1. First run the positive match stage
+        STAGE_MATCH_POSITIVE,
+        {
+            # 2. We then need to extract the date using substring since most dates look like this:
+            # "2020-05-10 07:30:00 UTC" but the `dateFromString` function does not handle the
+            # timezone string "UTC"
+            "$addFields": {
+                DATE_EXTRACTED: {
+                    "$substrBytes": [f"${FIELD_DATE_TESTED}", 0, 19],
+                },
+            },
         },
-    )
+        {
+            # 3. Then add a date field which converts the extracted date string to a mongo date
+            # type using the specified format. If it cannot parse the date using the first format
+            # it tries with the other format found in the data
+            # TODO: handle the case that no format is matched
+            "$addFields": {
+                DATE_CONVERTED: {
+                    "$dateFromString": {
+                        "dateString": f"${DATE_EXTRACTED}",
+                        "format": "%Y-%m-%d %H:%M:%S",
+                        "timezone": "UTC",
+                        "onError": {
+                            "$dateFromString": {
+                                "dateString": f"${DATE_EXTRACTED}",
+                                "format": "%d/%m/%Y %H:%M",
+                                "timezone": "UTC",
+                                "onError": DATE_ERROR,
```

Comment:
I'm thinking with this, we will never see in the report any sample that has a testing date not matching any of these 2 conditions. Do these 2 conditions match all testing dates in database? If that is the case, then is fine for the moment and we should replace this in future with a migration to use store this testing date as Date in that field (so the error on format gets captured before); but if we can have new data from plate maps that doesn't match these 2 conditions, then we could miss new samples in the report. Wouldn't it be easier in that case to use 'created_at' instead?

Reply:
Yes, I need to create an issue in the crawler for this.